sync2/migration.go

```diff
@@ -0,0 +1,370 @@
+	// handling here. Therefore this device is not useful to the proxy.
+	//
+	// If we leave this device's rows in situ, we may end up with rows in
+	// syncv3_to_device_messages, syncv3_to_device_ack_pos and syncv3_txns which have
+	// null values for the new fields, which will mean we fail to impose the uniqueness
+	// constraints at the end of the migration. Instead, drop those rows.
+	err = exec(
+		txn,
+		`DELETE FROM syncv3_sync2_devices WHERE device_id = $1`,
+		expectOneRowAffected,
+		device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`DELETE FROM syncv3_to_device_messages WHERE device_id = $1`,
+		expectAnyNumberOfRowsAffected,
+		device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`DELETE FROM syncv3_to_device_ack_pos WHERE device_id = $1`,
+		expectAtMostOneRowAffected,
+		device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`DELETE FROM syncv3_device_data WHERE device_id = $1`,
+		expectAtMostOneRowAffected,
+		device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	err = exec(
+		txn,
+		`DELETE FROM syncv3_txns WHERE user_id = $1`,
+		expectAnyNumberOfRowsAffected,
+		device.AccessTokenHash,
+	)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func exec(txn *sqlx.Tx, query string, checkRowsAffected func(ra int64) bool, args ...any) error {
+	res, err := txn.Exec(query, args...)
+	if err != nil {
+		return err
+	}
+	ra, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if !checkRowsAffected(ra) {
+		return fmt.Errorf("Failed checkRowsAffected: got %d", ra)
+	}
+	return nil
+}
+
+func expectOneRowAffected(ra int64) bool          { return ra == 1 }
+func expectAnyNumberOfRowsAffected(ra int64) bool { return true }
+func logRowsAffected(msg string) func(ra int64) bool {
+	return func(ra int64) bool {
+		logger.Info().Msgf(msg, ra)
+		return true
+	}
+}
+func expectAtMostOneRowAffected(ra int64) bool { return ra == 0 || ra == 1 }
+
+func finish(txn *sqlx.Tx) (err error) {
+	// OnExpiredToken used to delete from the to-device table, but not from the
+	// to-device ack pos table. Fix this up by deleting any orphaned ack pos rows.
+	err = exec(
+		txn,
+		`
+		DELETE FROM syncv3_to_device_ack_pos
+		WHERE device_id IN (
+		    SELECT syncv3_to_device_ack_pos.device_id
+		    FROM syncv3_to_device_ack_pos LEFT JOIN syncv3_sync2_devices USING (device_id)
+		);`,
+		logRowsAffected("Deleted %d stale rows from syncv3_to_device_ack_pos"),
+	)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_sync2_devices
+		DROP COLUMN v2_token_encrypted,
+		ADD PRIMARY KEY (user_id, device_id);
+	`)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_to_device_messages
+		ALTER COLUMN user_id SET NOT NULL;
+	`)
+	if err != nil {
+		return
+	}
+
+	_, err = txn.Exec(`
+		ALTER TABLE syncv3_to_device_ack_pos
+		ALTER COLUMN user_id SET NOT NULL,
```

Comment:
The cleanup for syncv3_to_device_ack_pos above ensures that this ALTER COLUMN succeeds.